[llvm-readobj, ELF] Support reading binary with more than PN_XNUM segments.

aokblast opened the pull request:
FreeBSD coredump uses program headers to store mmap information. It is possible for program to use more than PN_XNUM mmaps. Therefore, we implement the support of PN_XNUM in readelf.

jh7370 commented:
@aokblast, is this ready for review? It's currently still attempting to merge into another of your user branches, whereas if it's ready for review, I'd expect it to be merging into `main`.

aokblast committed:
[llvm-readobj, ELF] Support reading binary with more than PN_XNUM segments.

FreeBSD coredump uses program headers to store mmap information. It is possible
for program to use more than PN_XNUM mmaps. Therefore, we implement the support
of PN_XNUM in readelf.

jh7370 commented on llvm/tools/llvm-readobj/ELFDumper.cpp:
```diff
@@ -3572,45 +3572,62 @@ static inline void printFields(formatted_raw_ostream &OS, StringRef Str1,
+  Expected<uint32_t> PhNumOrErr = Obj.getPhNum();
+  if (!PhNumOrErr) {
+    // In this case we can ignore an error, because we have already reported a
+    // warning about the broken section header table earlier.
```
This feels fragile and it's also not the right use-case for `consumeError`. llvm-readobj has `reportUniqueWarning` to ensure the same warning isn't printed more than once.

jh7370 replied:
Please don't resolve comment threads that I have initiated.

jh7370 commented on llvm/tools/llvm-readobj/ELFDumper.cpp:
```diff
@@ -3572,45 +3572,62 @@ static inline void printFields(formatted_raw_ostream &OS, StringRef Str1,
-  if (ElfHeader.e_shnum != 0)
-    return to_string(ElfHeader.e_shnum);
+  if (Obj.getHeader().e_shnum != 0)
+    return to_string(Obj.getHeader().e_shnum);
```
Same comment as above.

jh7370 commented on llvm/tools/llvm-readobj/ELFDumper.cpp:
```diff
@@ -4778,8 +4795,10 @@ void GNUELFDumper<ELFT>::printProgramHeaders(
-    if (Header.e_phnum == 0) {
+    Expected<uint32_t> PhNumOrErr = this->Obj.getPhNum();
+    if (!PhNumOrErr) {
+      OS << '\n' << errorToErrorCode(PhNumOrErr.takeError()).message() << '\n';
```
This isn't how warnings are printed in llvm-readobj. Please review the existing code and fix it accordingly.

jh7370 commented on llvm/test/tools/llvm-readobj/ELF/invalid-e_phnum.test:
```diff
@@ -0,0 +1,39 @@
+
+# CASE-INVALID: SectionHeaderOffset: 0
+# CASE-INVALID: ProgramHeaderCount: 65535 (corrupt)
+# CASE-INVALID: unable to dump program headers: program headers are longer than binary of size 336: e_phoff = 0x40, e_phnum = 65535, e_phentsize = 56
```
This doesn't feel like the right warning to be printing. We should just be printing something about the section header table missing or similar. We should probably not continue to try to print the program header table, if the program header count cannot be determined due to an invalid section header table.

aokblast commented:
Sorry for the late reply, I give some feedback and please check it when you have time. Thanks!

jh7370 commented:
> Sorry for the late reply, I give some feedback and please check it when you have time. Thanks!

Ping, what's the situation here @aokblast? Last I know I made some comments that haven't been addressed.

aokblast committed:
fixup! [llvm-readobj, ELF] Support reading binary with more than PN_XNUM segments.

aokblast committed:
fixup! [llvm-readobj, ELF] Support reading binary with more than PN_XNUM segments.

jh7370 commented on llvm/test/tools/llvm-readobj/ELF/invalid-e_phnum.test:
```diff
@@ -0,0 +1,43 @@
+# CASE-INVALID: SectionHeaderOffset: 0x0
+# CASE-INVALID: ProgramHeaderEntrySize: 56
+# CASE-INVALID-NEXT: warning: '[[FILE]]': invalid section index: 0
+# CHECK-INVALID-NEXT:  ProgramHeaderCount: <?>
```
Inconsistent prefix names probably means these are being ignored.

aokblast commented:
ping @jh7370. Please take a look when you have time, thanks!

jh7370 commented on llvm/test/tools/llvm-readobj/ELF/many-sections.s:
```diff
@@ -37,12 +37,12 @@ Sections:
 
 # GNU2-EMPTY:
 # GNU2:       There are no sections in this file.
-# GNU2-NEXT:  warning: '[[FILE]]': e_shstrndx == SHN_XINDEX, but the section header table is empty
```
Is this warning no longer being emitted in favour of the earlier warning you've added a check for? If so, that feels suboptimal, as the new message is clearly more useful.

aokblast replied:
Sorry for the incorrect comment. The error is not reported for the following reason.

The message originates from:

```c++
if (*ShStrNdxOrErr == ELF::SHN_XINDEX && Sections.empty())
  return createError(
      "e_shstrndx == SHN_XINDEX, but the section header table is empty");
```

However, we already return an error earlier in:

```C++
if (!ShStrNdxOrErr)
  return ShStrNdxOrErr.takeError();
```

Since getShStrNdx has already handled and produced the initial error, we should either:


1. Propagate the error directly from getShStrNdx, or


2. Explicitly consume that error and then emit our own createError with the original message.


As written, the earlier return prevents the later createError from ever being reached.

jh7370 commented on llvm/test/tools/llvm-readobj/ELF/many-segments.test:
```diff
@@ -0,0 +1,64 @@
+# RUN: llvm-readobj --file-headers --sections --segments %t | FileCheck %s
+# RUN: llvm-readelf --segments %t | FileCheck --check-prefix=SYMS %s
+
+
```
Nit: remove second blank line

jh7370 reviewed (commented):
I've done a fairly quick whizz through this PR, but there are likely more issues that I haven't spotted, which we can come back to after you've addressed my current points.

aokblast committed:
fixup! [Object,ELF] Implement PN_XNUM extension for program headers (#162288)

jh7370 commented on llvm/include/llvm/Object/ELF.h:
```diff
@@ -935,9 +935,13 @@ template <class ELFT> ELFFile<ELFT>::ELFFile(StringRef Object) : Buf(Object) {}
+  // If e_shnum == 0 && e_shoff == 0, that means we actually have no section and
+  // thus is a valid ELF. But if the condition is still true when e_phnum ==
+  // PN_XNUM || e_shtstrndx == SHN_XINDEX and e_shoff == 0 The error will
+  // trigger when getSection() and e_shoff == 0
```
Hmm, it looks to me like you've changed the intent of this check, which is a bit outside the scope of this PR. Previously, if `e_shoff` was 0, we didn't emit an error for any of `e_phnum == ELF::PN_XNUM`, `e_shnum == 0`, or `e_shstrndx == ELF_SHN_XINDEX`. I accept that there's probably a case for emitting a warning in the e_phnum and e_shstrndx cases, but making that change of behaviour is independent of supporting reading more than PN_XNUM segments, which is what this PR is ostensibly about, so it should be in its own PR.

aokblast replied:
Grpahite tells me to rebase this PR since this one have already created and is not a descend of my new PR (ELF.h fix only).  Can I rebase it right now?

jh7370 commented on llvm/test/tools/llvm-readobj/ELF/many-segments.test:
```diff
@@ -0,0 +1,64 @@
+## Show that llvm-readelf can handle an input file with many segments.
+# RUN: yaml2obj %s -o %t
+# RUN: llvm-readobj --file-headers --sections --segments %t | FileCheck %s
```
But the contents of shdr 0 drives the interpretation of phnum, not the other way around, so testing the section header content doesn't give us anything.

Indeed, you don't even check the contents of the section header in any meaningful way...

aokblast replied:
Emm, I test the Sections [] in the readobj output. But you are right, it is meaningless and I decide to just check like what we have done in many-sections test.

aokblast committed:
[Object][ELF] Fix section header zero check

The PN_XUM is a necessary condition for reading shdr0
regardless of the value of e_shoff. Without this,
readShdrZero falsely returns the garbage value in ELF
header instead of emitting warning.

aokblast committed:
fixup! [Object][ELF] Fix section header zero check

jh7370 commented on llvm/test/tools/llvm-readobj/ELF/invalid-e_phnum.test:
```diff
@@ -0,0 +1,45 @@
+
+# RUN: llvm-readobj --headers %t2.o 2>&1 | FileCheck %s --check-prefix=CASE-VALID
+
+# CASE-VALID: SectionHeaderOffset: 0
```
Why are you testing this field?

jh7370 commented on llvm/test/tools/llvm-readobj/ELF/invalid-e_phnum.test:
```diff
@@ -0,0 +1,45 @@
+
+# RUN: yaml2obj --docnum=2 %s -o %t2.o
+
+# RUN: llvm-readobj --headers %t2.o 2>&1 | FileCheck %s --check-prefix=CASE-VALID
```
It's probably best to be explicit about which headers you're dumping - you want to dump both program headers and the file header, but you don't need other headers.

I'd also add `--implicit-check-not=warning:` to the FileCheck invocation, to show there are no other warnings produced.

Also, why is this case being tested in this file and not in `many-segments.test`?

jh7370 commented on llvm/test/tools/llvm-readobj/ELF/many-segments.test:
```diff
@@ -0,0 +1,29 @@
+# RUN: yaml2obj --docnum=1 %s -o %t1
+# RUN: llvm-readelf --headers %t1 | FileCheck %s --check-prefix=GNU
+## The ELF header should have e_phnum == PN_XNUM
+# GNU:         Number of program headers: 65535 (2)
```
Does GNU readelf have support for PN_XNUM and if so what is its output in this case?

aokblast replied:
```
x86_64-unknown-freebsd16.0-readelf --file-header /home/blast/Gits/llvm-project/build/test/tools/llvm-readobj/ELF/Output/many-segments.test.tmp1
```

Output:

```
  Number of program headers:         65535 (2)
```

jh7370 commented on llvm/test/tools/llvm-readobj/ELF/many-segments.test:
```diff
@@ -0,0 +1,29 @@
+    Info: 2
+ProgramHeaders:
+  - Type:     PT_LOAD
+    VAddr:    0x0001
```
The `VAddr` fields should be irrelevant and can be removed.

jh7370 commented on llvm/tools/llvm-readobj/ELFDumper.cpp:
```diff
@@ -4853,10 +4875,16 @@ template <class ELFT> void GNUELFDumper<ELFT>::printProgramHeaders() {
+  if (Expected<uint32_t> PhNumOrErr = this->Obj.getPhNum(); PhNumOrErr)
+    PhNum = *PhNumOrErr;
+  else
+    this->reportUniqueWarning(PhNumOrErr.takeError());
```
This isn't especially readable, in my opinion. I'd just follow what you do in the similar case in the function above.

I'd also look into what GNU does in this case, assuming GNU readelf supports PN_XNUM. It seems odd to report there are 0 program headers when the value is known to be incorrect.

aokblast committed:
fixup! [Object][ELF] Fix section header zero check

aokblast committed:
fixup! [Object][ELF] Fix section header zero check

aokblast commented on llvm/test/tools/llvm-readobj/ELF/many-sections.s:
```diff
@@ -37,6 +37,7 @@ Sections:
 # RUN:   FileCheck %s -DFILE=%t2 --check-prefix=GNU2
 # RUN: llvm-readelf --file-headers --section-details %t2 2>&1 | \
 # RUN:   FileCheck %s -DFILE=%t2 --check-prefix=GNU2
+# GNU2:  warning: '[[FILE]]': invalid section index: 0
```
Not reported anymore since https://github.com/llvm/llvm-project/pull/165278/changes#r2832084696.

aokblast committed:
fixup! [llvm-readobj, ELF] Support reading binary with more than PN_XNUM segments.

aokblast committed:
fixup! [Object][ELF] Fix section header zero check

aokblast committed:
fixup! [Object][ELF] Fix section header zero check

aokblast committed:
Merge branch 'users/aokblast/elf/fix_error' into users/aokblast/readelf/pxnum_support

aokblast committed:
fixup! Merge branch 'users/aokblast/elf/fix_error' into users/aokblast/readelf/pxnum_support

jh7370 commented on llvm/tools/llvm-readobj/ELFDumper.cpp:
```diff
@@ -4853,10 +4873,17 @@ template <class ELFT> void GNUELFDumper<ELFT>::printProgramHeaders() {
      << enumToString(Header.e_type, ArrayRef(ElfObjectFileType)) << "\n"
      << "Entry point " << format_hex(Header.e_entry, 3) << "\n"
-     << "There are " << Header.e_phnum << " program headers,"
+     << "There are " << PhNum << " program headers,"
```
I'd expect to see test coverage for this behaviour with `>= PN_XNUM` program headers. Also test coverage for this code executing when there's a problem with reading the program header count.

aokblast committed:
fixup! [Object][ELF] Fix section header zero check

aokblast committed:
fixup! [Object][ELF] Fix section header zero check

aokblast committed:
fixup! [llvm-readobj, ELF] Support reading binary with more than PN_XNUM segments.

aokblast committed:
Merge branch 'users/aokblast/elf/fix_error' into users/aokblast/readelf/pxnum_support

jh7370 commented on llvm/test/tools/llvm-readobj/ELF/file-headers.test:
```diff
@@ -222,14 +224,32 @@ Sections:
+# RUN:  | FileCheck %s -DFILE=%t.invalid1 -DSECHDROFF=4096 -DPHDRCOUNT="0" -DSECHDRCOUNT=8192 -DSECHDRSTRTABINDEX=12288 --check-prefix=INVALID-GNU
 
 ## Check what we print when e_shnum == 0, e_shstrndx == SHN_XINDEX and the section header table can't be read.
+## TODO: We should not report LLVM-BAD-PHDR and GNU-BAD-PHDR in the future since these are duplicate errors.
```
If we no longer get two warnings for the same issue, there's no need for this TODO any longer, is there? The problem before was that we were getting two warnings for the same test case, I believe.

aokblast committed:
fixup! [llvm-readobj, ELF] Support reading binary with more than PN_XNUM segments.

aokblast committed:
Merge branch 'main' into users/aokblast/readelf/pxnum_support

jh7370 commented on llvm/tools/llvm-readobj/ELFDumper.cpp:
Can this warning case actually be hit? I'd think that it's handled in the function immediately above. Are there other call sites to this function? If not, this case should be wrapped in a `cantFail`.

aokblast committed:
fixup! [llvm-readobj, ELF] Support reading binary with more than PN_XNUM segments.

jh7370 commented on llvm/tools/llvm-readobj/ELFDumper.cpp:
```diff
@@ -4897,10 +4898,12 @@ template <class ELFT> void GNUELFDumper<ELFT>::printProgramHeaders() {
     PhNum = *PhNumOrErr;
   else
-    this->reportUniqueWarning(PhNumOrErr.takeError());
+    llvm::consumeError(PhNumOrErr.takeError());
```
Does this actually ever get hit? As far as I can tell, the only case this function is called is from the other overload of `GNUELFDumper<ELFT>::printProgramHeaders`, but that code doesn't call this function, if `getPhNum` returns an error.

aokblast replied:
Yes, you are right. I misunderstand something, sorry. Switch to cantFail

jh7370 reviewed (commented):
I've been thinking more about the root of the problem of the near-duplicate warnings. The intention of `reportUniqueWarning` was to only print the same warning once during the run of the program, to reduce such duplication. This works fine for basic examples, but it breaks down when we add additional context like "unable to dump program headers" or "can't read program headers to build section to segment mapping". I think if we were to solve this problem, we could get rid of the duplicate warnings without needing to bend over backwards to make the rest of the code accommodate them.

This is out of scope for this PR, but one idea would be to change `reportUniqueWarning` to take a context prefix separate to the core body of the warning, then deduplicate the warning only on the body of the warning and not the prefix. This wouldn't work for every case (e.g. if we built up the context over multiple functions in the call stack), but it would be enough for the issues in this PR. What do you think?

aokblast committed:
fixup! [llvm-readobj, ELF] Support reading binary with more than PN_XNUM segments.

aokblast commented:
> I've been thinking more about the root of the problem of the near-duplicate warnings. The intention of `reportUniqueWarning` was to only print the same warning once during the run of the program, to reduce such duplication. This works fine for basic examples, but it breaks down when we add additional context like "unable to dump program headers" or "can't read program headers to build section to segment mapping". I think if we were to solve this problem, we could get rid of the duplicate warnings without needing to bend over backwards to make the rest of the code accommodate them.
> 
> This is out of scope for this PR, but one idea would be to change `reportUniqueWarning` to take a context prefix separate to the core body of the warning, then deduplicate the warning only on the body of the warning and not the prefix. This wouldn't work for every case (e.g. if we built up the context over multiple functions in the call stack), but it would be enough for the issues in this PR. What do you think?

I was thinking about this last week as well. The reason I didn’t include this idea here is due to your concern about rebuilding the context. But you are right, it might help us fix some duplications inside this PR. I can open a separate PR tomorrow. Thanks!

aokblast committed:
Merge branch 'main' into users/aokblast/readelf/pxnum_support

aokblast committed:
Merge branch 'main' into users/aokblast/readelf/pxnum_support

jh7370 reviewed (approved):
LGTM.

I'd still like a mechanism for avoiding the near-duplicate errors, but that's a follow-on piece of work and I don't have any further good ideas that are clean (perhaps changes to the dynamic table parsing as described would help, or adding an optional "context" argument to `reportUniqueWarning` would deal with the issue).

aokblast committed:
Merge branch 'main' into users/aokblast/readelf/pxnum_support

aokblast committed:
Merge branch 'main' into users/aokblast/readelf/pxnum_support

aokblast committed:
fixup! [llvm-readobj, ELF] Support reading binary with more than PN_XNUM segments.
